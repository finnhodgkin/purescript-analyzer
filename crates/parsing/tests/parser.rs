macro_rules! parser_tests {
    ($($name:ident => $source:expr),+ $(,)?) => {
        $(
            #[test]
            fn $name() {
                let lexed = lexing::lex($source);
                let tokens = lexing::layout(&lexed);
                let (node, errors) = parsing::parse(&lexed, &tokens);
                insta::assert_debug_snapshot!((node, errors));
            }
        )+
    };
}

parser_tests!(
    comment_pre_post => include_str!("parser/CommentPrePost.purs"),
    lexer_error_inserted => include_str!("parser/LexerErrorInserted.purs"),
    module_export => include_str!("parser/ModuleExport.purs"),
    module_export_class => include_str!("parser/ModuleExportClass.purs"),
    module_export_class_error_end => include_str!("parser/ModuleExportClassErrorEnd.purs"),
    module_export_comma_error => include_str!("parser/ModuleExportCommaError.purs"),
    module_export_data => include_str!("parser/ModuleExportData.purs"),
    module_export_data_comma_error => include_str!("parser/ModuleExportDataCommaError.purs"),
    module_export_data_item_after_double_period => include_str!("parser/ModuleExportDataItemAfterDoublePeriod.purs"),
    module_export_data_item_before_double_period => include_str!("parser/ModuleExportDataItemBeforeDoublePeriod.purs"),
    module_export_data_item_invalid => include_str!("parser/ModuleExportDataItemInvalid.purs"),
    module_export_empty => include_str!("parser/ModuleExportEmpty.purs"),
    module_export_invalid => include_str!("parser/ModuleExportInvalid.purs"),
    module_export_type_operator => include_str!("parser/ModuleExportTypeOperator.purs"),
    module_export_type_operator_error => include_str!("parser/ModuleExportTypeOperatorError.purs"),
    module_export_type_operator_error_end => include_str!("parser/ModuleExportTypeOperatorErrorEnd.purs"),
    module_header => include_str!("parser/ModuleHeader.purs"),
    module_header_prefixed => include_str!("parser/ModuleHeaderPrefixed.purs"),
);
