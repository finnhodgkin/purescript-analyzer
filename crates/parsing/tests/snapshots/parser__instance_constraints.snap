---
source: crates/parsing/tests/parser.rs
expression: "(node, errors)"
snapshot_kind: text
---
(
    Module@0..149
      ModuleHeader@0..34
        MODULE@0..6 "module"
        Comment@6..7
          WHITESPACE@6..7 " "
        ModuleName@7..27
          UPPER@7..26 "InstanceConstraints"
          Comment@26..27
            WHITESPACE@26..27 " "
        ModuleExportList@27..27
        WHERE@27..32 "where"
        Comment@32..34
          WHITESPACE@32..34 "\n\n"
      LAYOUT_START@34..34 ""
      ModuleImports@34..34
      ModuleStatements@34..149
        InstanceChain@34..149
          InstanceDeclaration@34..80
            INSTANCE@34..42 "instance"
            Comment@42..43
              WHITESPACE@42..43 " "
            InstanceConstraints@43..58
              TypeApplicationChain@43..55
                TypeConstructor@43..53
                  QualifiedName@43..53
                    UPPER@43..52 "Serialise"
                    Comment@52..53
                      WHITESPACE@52..53 " "
                TypeVariable@53..55
                  LOWER@53..54 "a"
                  Comment@54..55
                    WHITESPACE@54..55 " "
              RIGHT_THICK_ARROW@55..57 "=>"
              Comment@57..58
                WHITESPACE@57..58 " "
            InstanceHead@58..80
              UPPER@58..67 "Serialise"
              Comment@67..68
                WHITESPACE@67..68 " "
              TypeVariable@68..70
                LOWER@68..69 "s"
                Comment@69..70
                  WHITESPACE@69..70 " "
              TypeParenthesized@70..80
                LEFT_PARENTHESIS@70..71 "("
                TypeApplicationChain@71..78
                  TypeConstructor@71..77
                    QualifiedName@71..77
                      UPPER@71..76 "Array"
                      Comment@76..77
                        WHITESPACE@76..77 " "
                  TypeVariable@77..78
                    LOWER@77..78 "a"
                RIGHT_PARENTHESIS@78..79 ")"
                Comment@79..80
                  WHITESPACE@79..80 "\n"
          InstanceDeclaration@80..149
            ELSE@80..84 "else"
            Comment@84..85
              WHITESPACE@84..85 " "
            INSTANCE@85..93 "instance"
            Comment@93..94
              WHITESPACE@93..94 " "
            InstanceConstraints@94..124
              LEFT_PARENTHESIS@94..95 "("
              TypeApplicationChain@95..106
                TypeConstructor@95..105
                  QualifiedName@95..105
                    UPPER@95..104 "Serialise"
                    Comment@104..105
                      WHITESPACE@104..105 " "
                TypeVariable@105..106
                  LOWER@105..106 "a"
              COMMA@106..107 ","
              Comment@107..108
                WHITESPACE@107..108 " "
              TypeApplicationChain@108..119
                TypeConstructor@108..118
                  QualifiedName@108..118
                    UPPER@108..117 "Serialise"
                    Comment@117..118
                      WHITESPACE@117..118 " "
                TypeVariable@118..119
                  LOWER@118..119 "b"
              RIGHT_PARENTHESIS@119..120 ")"
              Comment@120..121
                WHITESPACE@120..121 " "
              RIGHT_THICK_ARROW@121..123 "=>"
              Comment@123..124
                WHITESPACE@123..124 " "
            InstanceHead@124..149
              UPPER@124..133 "Serialise"
              Comment@133..134
                WHITESPACE@133..134 " "
              TypeVariable@134..136
                LOWER@134..135 "s"
                Comment@135..136
                  WHITESPACE@135..136 " "
              TypeParenthesized@136..149
                LEFT_PARENTHESIS@136..137 "("
                TypeApplicationChain@137..147
                  TypeConstructor@137..144
                    QualifiedName@137..144
                      UPPER@137..143 "Either"
                      Comment@143..144
                        WHITESPACE@143..144 " "
                  TypeVariable@144..146
                    LOWER@144..145 "a"
                    Comment@145..146
                      WHITESPACE@145..146 " "
                  TypeVariable@146..147
                    LOWER@146..147 "b"
                RIGHT_PARENTHESIS@147..148 ")"
                Comment@148..149
                  WHITESPACE@148..149 "\n"
      LAYOUT_END@149..149 ""
      END_OF_FILE@149..149 ""
    ,
    [],
)
